### PR TITLE
[Merged by Bors] - feat(NumberTheory/AbelSummation): add more results

### PR DESCRIPTION
**Pull request by xroblot**
- Add versions for when the endpoints are `Nat` (for convenience).
- Add limit versions for when the endpoint tends to infinity.
- Add summability criteria using Abel summation

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by xroblot** [comment]
Question for reviewers: some results are stated in the case `c 0 = 0` but still, in this case, the statements contain terms like  `∑ k ∈ Icc 0 ⌊b⌋₊, f k * c k` even though the term `k = 0` does contribute to the sum.

It is better to leave it like that for consistency (and the user can remove the term `k  = 0` latter on if needed) or should I changed the statements so that summations start at `1`?

File: Mathlib/NumberTheory/AbelSummation.lean [hunk 1]
```diff
@@ -129,45 +159,217 @@ theorem _root_.sum_mul_eq_sub_sub_integral_mul (ha : 0 ≤ a) (hab : a ≤ b)
...
+  refine (locallyIntegrableOn_iff isLocallyClosed_Ici).mpr fun K hK₁ hK₂ ↦ ?_
+  by_cases hK₃ : K.Nonempty
+  · have h_inf : a ≤ sInf K := (hK₁ (hK₂.sInf_mem hK₃))
+    refine IntegrableOn.mono_set ?_ (Bornology.IsBounded.subset_Icc_sInf_sSup hK₂.isBounded)
+    refine integrablemulsum _ (ha.trans h_inf) (hg.mono_set ?_)
+    exact (Set.Icc_subset_Ici_iff (Real.sInf_le_sSup _ hK₂.bddBelow hK₂.bddAbove)).mpr h_inf
+  · rw [Set.not_nonempty_iff_eq_empty.mp hK₃]
+    exact integrableOn_empty
+
+theorem tendsto_sum_mul_atTop_eq_sub_integral (hf_diff : ∀ t ∈ Set.Ici 0, DifferentiableAt ℝ f t)
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/AbelSummation.lean:254]:
```suggestion
theorem tendsto_sum_mul_atTop_nhds_one_sub_integral (hf_diff : ∀ t ∈ Set.Ici 0, DifferentiableAt ℝ f t)
```
There is no equality in the statement. Same for the next theorem.

File: Mathlib/NumberTheory/AbelSummation.lean [hunk 2]
```diff
@@ -129,45 +159,217 @@ theorem _root_.sum_mul_eq_sub_sub_integral_mul (ha : 0 ≤ a) (hab : a ≤ b)
...
+    (h_lim: Tendsto (fun n : ℕ ↦ f n * ∑ k ∈ Icc 0 n, c k) atTop (𝓝 l))
+    {g : ℝ → ℝ} (hg_dom : (fun t ↦ deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k) =O[atTop] g)
+    (hg_int : IntegrableAtFilter g atTop) :
+    Tendsto (fun n : ℕ ↦ ∑ k ∈ Icc 0 n, f k * c k) atTop
+      (𝓝 (l - ∫ t in Set.Ioi 1, deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k)) := by
+  have h_lim' : Tendsto (fun n : ℕ ↦ ∫ t in Set.Ioc (1 : ℝ) n, deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k)
+      atTop (𝓝 (∫ t in Set.Ioi 1, deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k)) := by
+    refine Tendsto.congr' (by
+      filter_upwards [eventually_ge_atTop 1] with _ h
+      rw [← integral_of_le (Nat.one_le_cast.mpr h)]) ?_
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/AbelSummation.lean:283]: Here it is perhaps better to write
```suggestion
    refine Tendsto.congr' ?_ ?_
    · filter_upwards [eventually_ge_atTop 1] with _ h
      rw [← integral_of_le (Nat.one_le_cast.mpr h)]
```
The inlined tactic proof is a bit hard to parse.

File: Mathlib/NumberTheory/AbelSummation.lean [hunk 3]
```diff
@@ -129,45 +159,217 @@ theorem _root_.sum_mul_eq_sub_sub_integral_mul (ha : 0 ≤ a) (hab : a ≤ b)
...
+        · filter_upwards with t using norm_nonneg _
+
+theorem summable_mul_of_bigO_atTop₀ (hc : c 0 = 0)
+    (hf_diff : ∀ t ∈ Set.Ici 1, DifferentiableAt ℝ (fun x ↦ ‖f x‖) t)
+    (hf_int : IntegrableOn (deriv (fun t ↦ ‖f t‖)) (Set.Ici 1))
+    (h_bdd : (fun n : ℕ ↦ ‖f n‖ * ∑ k ∈ Icc 0 n, ‖c k‖) =O[atTop] fun _ ↦ (1 : ℝ))
+    {g : ℝ → ℝ} (hg₁ : (fun t ↦ deriv (fun t ↦ ‖f t‖) t * ∑ k ∈ Icc 0 ⌊t⌋₊, ‖c k‖) =O[atTop] g)
+    (hg₂ : IntegrableAtFilter g atTop) :
+    Summable (fun n : ℕ ↦ f n * c n) := by
+  obtain ⟨C₁, hC₁⟩ := Asymptotics.isBigO_one_nat_atTop_iff.mp h_bdd
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/AbelSummation.lean:343]: This proof is repeated essentially verbatim. Would it be possible to state a common generalization (e.g., working with `Ici k` and `c j = 0` for all `j < k` for arbitrary natural `k`) and then specialize it for the two relevant cases?

**Comment by MichaelStollBayreuth** [comment]
> Question for reviewers: some results are stated in the case `c 0 = 0` but still, in this case, the statements contain terms like `∑ k ∈ Icc 0 ⌊b⌋₊, f k * c k` even though the term `k = 0` does contribute to the sum.
> 
> It is better to leave it like that for consistency (and the user can remove the term `k = 0` latter on if needed) or should I changed the statements so that summations start at `1`?

I would say it is probably fine to leave it as it is for now. If it turns out that it would be convenient to have modified versions, they can still be added or changed.

File: Mathlib/NumberTheory/AbelSummation.lean (hunk 2 above)
Comment (xroblot) [reply Mathlib/NumberTheory/AbelSummation.lean]: I added another `have` before so that things get easier to parse hopefully.

File: Mathlib/NumberTheory/AbelSummation.lean (hunk 3 above)
Comment (xroblot) [reply Mathlib/NumberTheory/AbelSummation.lean]: Indeed. I moved the common parts into two auxiliary lemmas.

**Comment by xroblot** [comment]
> > Question for reviewers: some results are stated in the case `c 0 = 0` but still, in this case, the statements contain terms like `∑ k ∈ Icc 0 ⌊b⌋₊, f k * c k` even though the term `k = 0` does contribute to the sum.
> > It is better to leave it like that for consistency (and the user can remove the term `k = 0` latter on if needed) or should I changed the statements so that summations start at `1`?
> 
> I would say it is probably fine to leave it as it is for now. If it turns out that it would be convenient to have modified versions, they can still be added or changed.

Ok, thanks for your advice and the review.

**Review by MichaelStollBayreuth — APPROVED** [review]
LGTM now!
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge
